edb/edgeql/compiler/inference/cardinality.py

```diff
@@ -672,32 +672,65 @@ def __infer_func_call(
+
+    ret_lower_bound, ret_upper_bound = _card_to_bounds(return_card)
+
+    if ir.preserves_optionality:
+        # This is a generic aggregate function which preserves the
+        # optionality of its generic argument.  For simplicity we
+        # are deliberately not checking the parameters here as that
+        # would have been done at the time of declaration.
+        arg_cards = []
+
+        for arg in ir.args:
+            arg.cardinality = infer_cardinality(
+                arg.expr, scope_tree=scope_tree, ctx=ctx)
+            arg_cards.append(arg.cardinality)
+
+        arg_card = zip(*(_card_to_bounds(card) for card in arg_cards))
+        arg_lower, _ = arg_card
+        return _bounds_to_card(min(arg_lower), ret_upper_bound)
 
-    args = []
-    # process positional args
-    for arg, typemod in zip(ir.args, ir.params_typemods):
-        arg.cardinality = infer_cardinality(
-            arg.expr,
-            scope_tree=scope_tree, ctx=ctx)
-        if typemod is not SetOfType:
-            args.append(arg.expr)
-
-    # the cardinality of the function call depends on the cardinality
-    # of non-SET_OF arguments AND the cardinality of the function
-    # return value
-    if ir.typemod is SetOfType:
-        return MANY
     else:
-        if args:
-            return _common_cardinality(
-                args, scope_tree=scope_tree, ctx=ctx,
-            )
-        else:
-            if ir.typemod is qltypes.TypeModifier.OptionalType:
-                return AT_MOST_ONE
+        # For regular non-OPTIONAL functions, the general rule of
+        # Cartesian cardinality of argument applies, although we still
+        # have to account for the declared return cardinality, as the
+        # function might be OPTIONAL or SET OF in its return type.
+        #
+        # If a function is OPTIONAL in its parameters, which includes
+        # aggregate functions, then we compute a Cartesian cardinality
+        # of functions's _non-OPTIONAL_ arguments and its return
+        # cardinality, but only in the upper bound, since we cannot know
+        # how the function behaves in OPTIONAL arguments.
+        singleton_args = []
+        singleton_arg_cards = []
+        all_singletons = True
+
+        for arg, typemod in zip(ir.args, ir.params_typemods):
+            arg.cardinality = infer_cardinality(
+                arg.expr, scope_tree=scope_tree, ctx=ctx)
+            if typemod is qltypes.TypeModifier.SingletonType:
+                singleton_args.append(arg.expr)
+                singleton_arg_cards.append(arg.cardinality)
             else:
-                return ONE
+                all_singletons = False
+
+        if not singleton_args:
+            # Either no arguments at all, or all arguments are non-singletons,
+            # so the declared return cardinality is as specific as we can get.
+            return return_card
+        else:
+            _check_op_volatility(singleton_args, singleton_arg_cards, ctx=ctx)
```

Comment:
I think we need to do `_check_op_volatility` on at least the OPTIONAL arguments. (Though I'd probably do it for SET OF ones too, unless we find an actual use-case and are convinced it works?)

Reply:
I'm not sure `_check_op_volatility` would work correctly for `SET OF` args.  Consider this hypothetical function:

```
func string_agg(items: SET OF str, delimiter: str) -> str { ... } 
```

and then

```SELECT string_agg({"foo", "bar"}, <str>random())```

This is a perfectly legal call, but since `items` is `MANY`, `_check_op_volatility` would trip, despite no Cartesian product actually being made.